core/src/test/scala/unit/kafka/utils/TestUtils.scala

```diff
@@ -1127,6 +1127,23 @@ object TestUtils extends Logging {
       throw new IllegalStateException(s"Cannot get topic: $topic, partition: $partition in server metadata cache"))
   }
 
+  /**
+   * Wait until the kraft broker metadata have caught up to the controller
+   */
+  def waitForKRaftBrokerMetadataCatchupController(
+      brokers: Seq[KafkaBroker],
+      controllerServer: ControllerServer,
+      msg: String = "Timeout waiting for controller metadata propagating to brokers"
+  ): Unit = {
+    TestUtils.waitUntilTrue(
+      () => {
+        brokers.forall { broker =>
+          val metadataOffset = broker.metadataCache.asInstanceOf[KRaftMetadataCache].currentImage().highestOffsetAndEpoch().offset
+          metadataOffset >= controllerServer.raftManager.replicatedLog.endOffset().offset - 1
```

Comment:
Would it make sense to fix the end offset at the start so that all brokers are catching up to the same point?

Reply:
Yeah, that makes sense, I also added a comment to this method that we should make sure the related metadata message has already been committed to the controller metadata log.